server/svix-server/src/worker.rs

```diff
@@ -291,6 +257,117 @@ fn bytes_to_string(bytes: bytes::Bytes) -> String {
+        cache.clone(),
+        db,
+        None,
+        msg_clone.app_id,
+        msg_clone.org_id,
+        Duration::from_secs(30),
+    )
+    .await?
+    .ok_or_else(|| Error::Generic(format!("Application doesn't exist: {}", &msg.app_id)))?;
+
+    let endpoints: Vec<CreateMessageEndpoint> = create_message_app
+        .endpoints
+        .iter()
+        .filter(|endpoint| match &queue_task {
+            QueueTask::MessageV1(task) => task.endpoint_id == endpoint.id,
+            QueueTask::MessageBatch(_) => true
+        })
+        .filter(|endpoint| {
+            return
+            // No disabled or deleted endpoints ever
+               !endpoint.disabled && !endpoint.deleted &&
+            (
+                // Manual attempt types go through regardless
+                queue_task.clone().trigger_type() == MessageAttemptTriggerType::Manual
+                || (
+                        // If an endpoint has event types and it matches ours, or has no event types
+                        endpoint
+                        .event_types_ids
+                        .as_ref()
+                        .map(|x| x.0.contains(&msg.event_type))
+                        .unwrap_or(true)
+                    &&
+                        // If an endpoint has no channels accept all messages, otherwise only if their channels overlap.
+                        // A message with no channels doesn't match an endpoint with channels.
+                        endpoint
+                        .channels
+                        .as_ref()
+                        .map(|x| !x.0.is_disjoint(msg.channels.as_ref().map(|x| &x.0).unwrap_or(&HashSet::new())))
+                        .unwrap_or(true)
+            ))})
+        .cloned()
+        .collect();
+
+    // TODO: remove this section once destinations are obsolete
+    let destinations = messagedestination::Entity::secure_find_by_msg(queue_task.clone().msg_id())
+        .all(db)
+        .await
+        .unwrap();
+    if destinations.is_empty() {
+        let destinations = endpoints
+            .iter()
+            .map(|endpoint| messagedestination::ActiveModel {
+                msg_id: Set(msg.id.clone()),
+                endp_id: Set(endpoint.id.clone()),
+                next_attempt: Set(Some(Utc::now().into())),
+                status: Set(MessageStatus::Sending),
+                ..Default::default()
+            });
+        messagedestination::Entity::insert_many(destinations)
+            .exec(db)
+            .await?;
+    }
+
+    let futures: Vec<_> = endpoints
+        .iter()
+        .map(|endpoint| {
+            dispatch(
+                cfg.clone(),
+                db,
+                queue_tx,
+                payload.clone(),
+                queue_task.clone().to_msg_task(endpoint.clone().id),
+                endpoint.to_owned(),
+            )
+        })
+        .collect();
+
+    let join = future::join_all(futures).await;
```

Comment:
> If any future returns an error then all other futures will be canceled and an error will be returned immediately.

From the docs for join_all. This is not good, it should not cancel others because it'll just put us in a weird state unnecessarily. I think @svix-james also mentioned it before.

Reply:
I think this is for [`try_join_all`](https://docs.rs/futures/0.3.21/futures/future/fn.try_join_all.html), which we moved away from following @svix-james's comment.

Reply:
Wasn't this @svix-james's to *not* cancel them prematurely?

Reply:
> If any future returns an error then all other futures will be canceled and an error will be returned immediately.

This is from the docs for `try_join_all`. `join_all` does not cancel prematurely, and I have to manually check the returned vector for error types (done below this line).

Reply:
Man, I literally copy-pasted it from the doc, I don't see how I could have looked at the wrong place, but it looks like I have. 🤷🏻